api/internal/store/infra/store.go

```diff
@@ -6,73 +6,100 @@ import (
 	"time"
 
 	"github.com/replicatedhq/embedded-cluster/api/types"
+	"github.com/tiendc/go-deepcopy"
 )
 
 const maxLogSize = 100 * 1024 // 100KB total log size
 
+var _ Store = &MemoryStore{}
+
 // Store provides methods for storing and retrieving infrastructure state
 type Store interface {
-	Get() (*types.Infra, error)
-	GetStatus() (*types.Status, error)
+	Get() (types.Infra, error)
+	GetStatus() (types.Status, error)
 	SetStatus(status types.Status) error
 	SetStatusDesc(desc string) error
 	RegisterComponent(name string) error
-	SetComponentStatus(name string, status *types.Status) error
+	SetComponentStatus(name string, status types.Status) error
 	AddLogs(logs string) error
 	GetLogs() (string, error)
 }
 
-// memoryStore is an in-memory implementation of Store
-type memoryStore struct {
-	infra *types.Infra
+// MemoryStore is an in-memory implementation of Store
+type MemoryStore struct {
+	infra types.Infra
 	mu    sync.RWMutex
 }
 
+type StoreOption func(*MemoryStore)
+
+func WithInfra(infra types.Infra) StoreOption {
+	return func(s *MemoryStore) {
+		s.infra = infra
+	}
+}
+
 // NewMemoryStore creates a new memory store
-func NewMemoryStore(infra *types.Infra) Store {
-	return &memoryStore{
-		infra: infra,
+func NewMemoryStore(opts ...StoreOption) Store {
+	s := &MemoryStore{}
+
+	for _, opt := range opts {
+		opt(s)
 	}
+
+	return s
 }
 
-func (s *memoryStore) Get() (*types.Infra, error) {
+func (s *MemoryStore) Get() (types.Infra, error) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return s.infra, nil
+
+	var infra types.Infra
+	if err := deepcopy.Copy(&infra, &s.infra); err != nil {
+		return types.Infra{}, err
+	}
+
+	return infra, nil
 }
 
-func (s *memoryStore) GetStatus() (*types.Status, error) {
+func (s *MemoryStore) GetStatus() (types.Status, error) {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	return s.infra.Status, nil
+
+	var status types.Status
+	if err := deepcopy.Copy(&status, &s.infra.Status); err != nil {
+		return types.Status{}, err
+	}
+
+	return status, nil
 }
 
-func (s *memoryStore) SetStatus(status types.Status) error {
+func (s *MemoryStore) SetStatus(status types.Status) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	s.infra.Status = &status
+	s.infra.Status = status
 	return nil
 }
 
-func (s *memoryStore) SetStatusDesc(desc string) error {
+func (s *MemoryStore) SetStatusDesc(desc string) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	if s.infra.Status == nil {
-		return fmt.Errorf("status not set")
+	if s.infra.Status.State == "" {
+		return fmt.Errorf("state not set")
 	}
```

Comment:
this is a little weird. any reason not to keep things as pointers on the interface and still return deep copies?

Reply:
it checked if status was nil, not if state was an empty string. this sets description not state.